ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -956,7 +961,12 @@ boolean getIndexEntriesSince(StoreKey key, FindEntriesCondition findEntriesCondi
    * @param entries the entries to eliminate duplicates from.
    */
   private void eliminateDuplicates(List<IndexEntry> entries) {
-    Set<StoreKey> setToFindDuplicate = new HashSet<StoreKey>();
+    Set<StoreKey> setToFindDuplicate = new HashSet<>();
+    // first choose PUTs over update entries (omitting DELETEs)
+    entries.removeIf(
+        entry -> !entry.getValue().isFlagSet(IndexValue.Flags.Delete_Index) && !setToFindDuplicate.add(entry.getKey()));
+    // then choose DELETEs over all other entries
+    setToFindDuplicate.clear();
     ListIterator<IndexEntry> iterator = entries.listIterator(entries.size());
     while (iterator.hasPrevious()) {
       IndexEntry entry = iterator.previous();
```

Comment:
`Keep-only-the-first-entry` (first being the last with reverse iteration), can be written without streams as follows:
```java
if (!setToFindDuplicate.add(entry.getKey()) {
  iterator.remove();
}
```

Reply:
Done